ext/pdo_sqlite/pdo_sqlite_class.c

```diff
@@ -0,0 +1,657 @@
+				stream->eof = 0;
+				return 0;
+			}
+		case SEEK_END:
+			if (offset > 0) {
+				sqlite3_stream->position = sqlite3_stream->size;
+				*newoffs = -1;
+				return -1;
+			} else if (sqlite3_stream->size < (size_t)(-offset)) {
+				sqlite3_stream->position = 0;
+				*newoffs = -1;
+				return -1;
+			} else {
+				sqlite3_stream->position = sqlite3_stream->size + offset;
+				*newoffs = sqlite3_stream->position;
+				stream->eof = 0;
+				return 0;
+			}
+		default:
+			*newoffs = sqlite3_stream->position;
+			return -1;
+	}
+}
+/* }}} */
+
+
+static int php_pdosqlite3_stream_cast(php_stream *stream, int castas, void **ret)
+{
+	return FAILURE;
+}
+
+static int php_pdosqlite3_stream_stat(php_stream *stream, php_stream_statbuf *ssb)
+{
+	php_stream_pdosqlite3_data *sqlite3_stream = (php_stream_pdosqlite3_data *) stream->abstract;
+	ssb->sb.st_size = sqlite3_stream->size;
+	return 0;
+}
+
+static const php_stream_ops php_stream_pdosqlite3_ops = {
+	php_pdosqlite3_stream_write,
+	php_pdosqlite3_stream_read,
+	php_pdosqlite3_stream_close,
+	php_pdosqlite3_stream_flush,
+	"PDOSQLite",
+	php_pdosqlite3_stream_seek,
+	php_pdosqlite3_stream_cast,
+	php_pdosqlite3_stream_stat,
+	NULL
+};
+
+
+/* {{{ Open a blob as a stream which we can read / write to. */
+PHP_METHOD(PdoSqlite, openBlob)
+{
+	char *table, *column, *dbname = "main", *mode = "rb";
+	size_t table_len, column_len, dbname_len;
+	zend_long rowid, flags = SQLITE_OPEN_READONLY, sqlite_flags = 0;
+	sqlite3_blob *blob = NULL;
+	php_stream_pdosqlite3_data *sqlite3_stream;
+	php_stream *stream;
+
+	pdo_dbh_t *dbh;
+	pdo_sqlite_db_handle *db_handle;
+
+	dbh = Z_PDO_DBH_P(ZEND_THIS);
+	PDO_CONSTRUCT_CHECK;
+	db_handle = (pdo_sqlite_db_handle *)dbh->driver_data;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "ssl|pl", &table, &table_len, &column, &column_len, &rowid, &dbname, &dbname_len, &flags) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	sqlite3 *sqlite_handle;
+	sqlite_handle = db_handle->db;
+	sqlite_flags = (flags & SQLITE_OPEN_READWRITE) ? 1 : 0;
+
+	if (sqlite3_blob_open(sqlite_handle, dbname, table, column, rowid, sqlite_flags, &blob) != SQLITE_OK) {
+	// TODO - exception needed here.
+//		php_sqlite3_error(db_obj, "Unable to open blob: %s", sqlite3_errmsg(db_obj->db));
+		// and converting to exception zend_throw_exception_ex(php_pdo_get_exception(), 0, errtext);
+		RETURN_FALSE;
+	}
+
+	sqlite3_stream = emalloc(sizeof(php_stream_pdosqlite3_data));
+	sqlite3_stream->blob = blob;
+	sqlite3_stream->flags = flags;
+	sqlite3_stream->position = 0;
+	sqlite3_stream->size = sqlite3_blob_bytes(blob);
+
+	if (sqlite_flags != 0) {
+		mode = "r+b";
+	}
+
+	stream = php_stream_alloc(&php_stream_pdosqlite3_ops, sqlite3_stream, 0, mode);
+
+	if (stream) {
+		php_stream_to_zval(stream, return_value);
+	} else {
+		RETURN_FALSE;
+	}
+}
+/* }}} */
+
+static int php_pgsql_collation_callback(void *context,
+	int string1_len, const void *string1,
+	int string2_len, const void *string2)
+{
```

Comment:
``pgsql`` naming, also is this a function pointer for SQLite3 which has the weird types for the lengths and string?

Reply:
Naming fixed, code was copied from pdo_sqlite particularly the hacky method that magically appears if pdo connects to SQLite, so I'm assuming the parameter types are correct.